modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermarkImpl.java

```diff
@@ -0,0 +1,280 @@
+
+    private final HybridClock clock;
+
+    private final TxManager txManager;
+
+    private final VaultManager vaultManager;
+
+    private final List<LowWatermarkChangedListener> updateListeners = new CopyOnWriteArrayList<>();
+
+    private final ScheduledExecutorService scheduledThreadPool;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    private volatile @Nullable HybridTimestamp lowWatermark;
+
+    private final AtomicReference<ScheduledFuture<?>> lastScheduledTaskFuture = new AtomicReference<>();
+
+    private final FailureProcessor failureProcessor;
+
+    /**
+     * Constructor.
+     *
+     * @param nodeName Node name.
+     * @param lowWatermarkConfig Low watermark configuration.
+     * @param clock A hybrid logical clock.
+     * @param txManager Transaction manager.
+     * @param vaultManager Vault manager.
+     * @param failureProcessor Failure processor tha is used to handle critical errors.
+     */
+    public LowWatermarkImpl(
+            String nodeName,
+            LowWatermarkConfiguration lowWatermarkConfig,
+            HybridClock clock,
+            TxManager txManager,
+            VaultManager vaultManager,
+            FailureProcessor failureProcessor
+    ) {
+        this.lowWatermarkConfig = lowWatermarkConfig;
+        this.clock = clock;
+        this.txManager = txManager;
+        this.vaultManager = vaultManager;
+        this.failureProcessor = failureProcessor;
+
+        scheduledThreadPool = Executors.newSingleThreadScheduledExecutor(
+                NamedThreadFactory.create(nodeName, "low-watermark-updater", LOG)
+        );
+    }
+
+    /**
+     * Starts the watermark manager.
+     */
+    @Override
+    public CompletableFuture<Void> start() {
+        inBusyLock(busyLock, () -> {
+            lowWatermark = readLowWatermarkFromVault();
+        });
+
+        return nullCompletedFuture();
+    }
+
+    /**
+     * Schedule watermark updates.
+     */
+    public void scheduleUpdates() {
+        inBusyLock(busyLock, () -> {
+            HybridTimestamp lowWatermarkCandidate = lowWatermark;
+
+            if (lowWatermarkCandidate == null) {
+                LOG.info("Previous value of the low watermark was not found, will schedule to update it");
+
+                scheduleUpdateLowWatermarkBusy();
+
+                return;
+            }
+
+            LOG.info("Low watermark has been scheduled to be updated: {}", lowWatermarkCandidate);
+
+            txManager.updateLowWatermark(lowWatermarkCandidate)
+                    .thenComposeAsync(unused -> inBusyLock(busyLock, () -> notifyListeners(lowWatermarkCandidate)), scheduledThreadPool)
+                    .whenComplete((unused, throwable) -> {
+                        if (throwable == null) {
+                            inBusyLock(busyLock, this::scheduleUpdateLowWatermarkBusy);
```

Comment:
Why don't we save the new low watermark in vault?

Reply:
This method a recovery method and it's is called on start.
Here we've just read watermark from the vault and applying it.

Reply:
I looked at the code again, maybe this code is not needed, maybe in this method it will be enough to schedule low watermark updates and that’s it? Because each component can get a low watermark at the start and do with it whatever they want and need. What do you think?

Reply:
That's too much refactoring non-relating to this ticket.
It would be better to create a separate ticket If you have any concerns.

Reply:
Okay let's create it.
This is one of the reasons to divide the task into several small ones.

Reply:
https://issues.apache.org/jira/browse/IGNITE-21751